File: decorator_pattern/dp.py
class Coffee : 
  def cost(self):
    return 5
  
class CoffeDecorator(Coffee):
  def __init__(self, coffee):
    self._coffee = coffee
  def cost(self):
    return self._coffee.cost()
  
class MilkDecorator(CoffeDecorator):
  def cost(self):
    return self._coffee.cost() + 2

class SugarDecorator(CoffeDecorator):
  def cost(self):
    return self._coffee.cost() + 1

File: decorator_pattern/dp.kt
open class Coffee {
  open fun cost(): Int {
    return 5
  }
}

open class CoffeeDecorator(private val coffee: Coffee) : Coffee() {
  override fun cost(): Int {
    return coffee.cost()
  }
}

class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
  override fun cost(): Int {
    return super.cost() + 2
  }
}

class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
  override fun cost(): Int {
    return super.cost() + 1
  }
}